pyvista/core/validation/_cast_array.py

```diff
@@ -0,0 +1,116 @@
+"""Array casting functions."""
+
+import numpy as np
+
+
+def _cast_to_list(arr):
+    """Cast an array to a nested list.
+
+    Parameters
+    ----------
+    arr : array_like
+        Array to cast.
+
+    Returns
+    -------
+    list
+        List or nested list array.
+    """
+    return _cast_to_numpy(arr).tolist()
+
+
+def _cast_to_tuple(arr):
+    """Cast an array to a nested tuple.
+
+    Parameters
+    ----------
+    arr : array_like
+        Array to cast.
+
+    Returns
+    -------
+    tuple
+        Tuple or nested tuple array.
+    """
+    arr = _cast_to_numpy(arr).tolist()
+
+    def _to_tuple(s):
+        return tuple(_to_tuple(i) for i in s) if isinstance(s, list) else s
+
+    return _to_tuple(arr)
+
+
+def _cast_to_numpy(arr, /, *, as_any=True, dtype=None, copy=False, must_be_real=False):
+    """Cast array to a NumPy ndarray.
+
+    Object arrays are not allowed but the dtype is otherwise unchecked by default.
+    String arrays and complex numbers are therefore allowed.
+
+    .. warning::
+
+        Arrays intended for use with vtk should set ``must_be_real=True``
+        since ``numpy_to_vtk`` uses the array values directly without
+        checking for complex arrays.
+
+    Parameters
+    ----------
+    arr : array_like
+        Array to cast.
+
+    as_any : bool, default: True
+        Allow subclasses of ``np.ndarray`` to pass through without
+        making a copy.
+
+    dtype : dtype_like, optional
+        The data-type of the returned array.
+
+    copy : bool, default: False
+        If ``True``, a copy of the array is returned. A copy is always
+        returned if the array:
+
+            * is a nested sequence
+            * is a subclass of ``np.ndarray`` and ``as_any`` is ``False``.
+
+    must_be_real : bool, default: True
+        Raise a TypeError if the array does not have real numbers, i.e.
+        its data type is not integer or floating.
+
+    Raises
+    ------
+    ValueError
+        If input cannot be cast as a NumPy ndarray.
+        ValueError
```

Comment:
`ValueError` duplicated here

Reply:
I see, that part of the docs is from the original function. This emulates NumPy's behaviour for ragged arrays, e.g. a ValueError is raised for `np.array([[1,2],[4,5,6]])`.

Just below the reviewed code block is new docs added for the TypeError generated if `must_be_real=True`.

But, the original function also forbids object dtype and also raises a ValueError there.

I think any dtype errors, either object or not real numbers, should be TypeErrors.

Reply:
I just mean the term `ValueError` is listed twice instead of once.

> I think any dtype errors, either object or not real numbers, should be TypeErrors.

Yeah, that makes sense to me.

Reply:
Oh I see, sorry I thought it was related to your other comment. I'll fix that too.